### PR TITLE
Fix Erlang code generation for strings with as pattern

### LINKED ISSUE #3250: Compiler mishandling case statements
I'm running Gleam 1.2.0 on Linux. 

The following code compiles into erlang, but the erlang code is invalid:

```gleam
import gleam/io

pub fn main() {
  let wibble = "wibble"
  let wobble = "wobble"

  // Multi-case is important here, I couldn't trigger this on a single input.
  case wibble, wobble {
    _, "\t" as space <> rest | _, " " as space <> rest ->
      io.debug(#(space, rest))
    _, _ -> io.debug(#("Nothing matched", wibble))
  }
}
```

```erlang
-module(bug_repro).
-compile([no_auto_import, nowarn_unused_vars, nowarn_unused_function, nowarn_nomatch]).

-export([main/0]).

-spec main() -> {binary(), binary()}.
main() ->
    Wibble = <<"wibble"/utf8>>,
    Wobble = <<"wobble"/utf8>>,
    case {Wibble, Wobble} of
        {_, <<Space:1/binary, Rest/binary>> when Space =:= <<"\t"/utf8>>} ->
            gleam@io:debug({Space, Rest});

        {_, <<Space:1/binary, Rest/binary>> when Space =:= <<" "/utf8>>} ->
            gleam@io:debug({Space, Rest});

        {_, _} ->
            gleam@io:debug({<<"Nothing matched"/utf8>>, Wibble})
    end.
```

Compiler output:

```
~/build/dev/erlang/commonmark/_gleam_artefacts/bug_repro.erl:11:45: syntax error before: 'when'
%   11|         {_, <<Space:1/binary, Rest/binary>> when Space =:= <<"\t"/utf8>>} ->
%     |                                             ^

~/build/dev/erlang/commonmark/_gleam_artefacts/bug_repro.erl:4:2: function main/0 undefined
%    4| -export([main/0]).
%     |  ^

~/build/dev/erlang/commonmark/_gleam_artefacts/bug_repro.erl:6:2: spec for undefined function main/0
%    6| -spec main() -> {binary(), binary()}.
%     |  ^
```

### PR DESCRIPTION
**Pull request by giacomocavalieri**
This PR closes #3250

### CODE REVIEW

**Comment by inoas** [comment]
Thank you Jak <3

File: CHANGELOG.md [hunk 1]
```diff
@@ -127,6 +127,10 @@
   documentation when they contained lines 79 or 80 characters long.
   ([Richard Viney](https://github.com/richard-viney))
 
+- Fixed a bug where the compiler would generate invalid erlang when pattern
```
Comment (lpil) [inline CHANGELOG.md:130]:
```suggestion
- Fixed a bug where the compiler would generate invalid Erlang when pattern
```

**Review by lpil — APPROVED** [review]
Lovely, thank you

**Comment by giacomocavalieri** [comment]
Done! Thank you

**Review by lpil — APPROVED** [review]
Thank you!

**Pull request merged**